atlasdb-feign/src/test/java/com/palantir/atlasdb/http/ExceptionCountingRefreshingClientTest.java

```diff
@@ -0,0 +1,148 @@
+
+import feign.Client;
+import feign.Request;
+import feign.Response;
+
+public class ExceptionCountingRefreshingClientTest {
+    private Supplier<Client> clientSupplier = (Supplier<Client>) mock(Supplier.class);
+    private Client client = mock(Client.class);
+
+    private Request request = Request.create("POST", "atlas.palantir.pt", ImmutableMap.of(), null, null);
+    private Response response = Response.create(204, "no content", ImmutableMap.of(), new byte[0]);
+    private Request.Options options = new Request.Options(1, 1);
+
+    private static final int EXCEPTION_COUNT_BEFORE_REFRESH = 2;
+    private static final RuntimeException CLIENT_EXCEPTION = new RuntimeException("bad");
+
+    @Test
+    public void passesRequestsThroughToTheDelegate() throws IOException {
+        when(clientSupplier.get()).thenReturn(client);
+        when(client.execute(request, options)).thenReturn(response);
+
+        Client refreshingClient = ExceptionCountingRefreshingClient.createRefreshingClient(clientSupplier);
+        callExecute(refreshingClient, 1);
+
+        verify(client, times(1)).execute(request, options);
+        verifyNoMoreInteractions(client);
+    }
+
+    @Test
+    public void createsClientOnlyOnceIfMakingMultipleRequests() throws IOException {
+        when(clientSupplier.get()).thenReturn(client);
+        when(client.execute(request, options)).thenReturn(response);
+
+        Client refreshingClient = ExceptionCountingRefreshingClient.createRefreshingClient(clientSupplier);
+        callExecute(refreshingClient, 3);
+
+        verify(clientSupplier, times(1)).get();
+        verify(client, times(3)).execute(request, options);
+        verifyNoMoreInteractions(clientSupplier, client);
+    }
+
+    @Test
+    public void reinvokesSupplierAfterTheSpecifiedNumberOfExceptionsThrown() throws IOException {
+        when(clientSupplier.get()).thenReturn(client);
+        when(client.execute(request, options)).thenThrow(CLIENT_EXCEPTION);
+
+        Client refreshingClient = new ExceptionCountingRefreshingClient(clientSupplier, EXCEPTION_COUNT_BEFORE_REFRESH);
+
+        int numberOfExecutions = EXCEPTION_COUNT_BEFORE_REFRESH + 1;
+        callExecute(refreshingClient, numberOfExecutions);
+
+        verify(clientSupplier, times((numberOfExecutions / EXCEPTION_COUNT_BEFORE_REFRESH + 1))).get();
+        verify(client, times(numberOfExecutions)).execute(request, options);
+        verifyNoMoreInteractions(clientSupplier, client);
+    }
+
+    @Test
+    public void requestsContinueWithOldClientIfDelegateSupplierThrows() throws IOException {
+        when(clientSupplier.get()).thenReturn(client);
+        when(client.execute(request, options)).thenThrow(CLIENT_EXCEPTION);
+
+        Client refreshingClient = new ExceptionCountingRefreshingClient(clientSupplier, EXCEPTION_COUNT_BEFORE_REFRESH);
+        callExecute(refreshingClient,1);
+
+        when(clientSupplier.get()).thenThrow(new IllegalStateException("bad"));
```

Comment:
I _think_ this is equivalent to writing one line at the beginning `when(clientSupplier.get()).thenReturn(client).thenThrow(new ISE("bad"));`. This could help us avoid having to do `numberOfExecutions - 1` calls in line 98, which requires looking backwards to figure out why there's a `- 1` there.